### PR TITLE
Iterative decoding with annotations

### PR DESCRIPTION
**Pull request by luanaruiz9**
**Updated data_generation.py to output number of ops:** data_generation.py now produces an .ops file for the it_dec_test split containing the number of operations in each input sequence of the iterative decoding test split. The number of operations is used to inform the number of prediction loops necessary to decode the iterative decoding test samples.

**Updated train.py to use annotations in iterative decoding:** see config.use_annotations. Other changes include: including the possibility to add extra prediction loops (to the number of loops obtained from the data), see config.extra_loops; printing wrong predictions instead of random predictions; saving list of predictions in a pickle file; adding the possibility of using a saved model to only perform predictions, see config.just_do_pred.

**Updated default.py:** to include flags use_annotations (use annotated number of operations to inform the number of prediction loops at test time), extra_loops (number of loops in addition to those obtained from annotations) and just_do_pred (only do predictions from a saved model).

**Updated input_pipeline.py to handle annotations**

### CODE REVIEW

File: datasets/pcfg/data_generation.py [hunk 1]
```diff
@@ -149,26 +148,31 @@ def main(argv: Sequence[str]):
     
     out_file_src = open(data_path + 'it_dec_val.src', 'w')
     out_file_tgt = open(data_path + 'it_dec_val.tgt', 'w')
```
Comment (santiaontanon-google) [inline datasets/pcfg/data_generation.py]: Sorry for not catching this earlier. I Was struggling to understand the code below, since I did not understand what "out_file_src_dec" stood for. If I understand correctly "out_file_src" is for the "validation set" and "out_file_src_dec" for the test set, right? Maybe we should call them "out_file_src_validation" and "out_file_src_test" instead to make it more readable? (same with *_tgt)

File: datasets/pcfg/data_generation.py [hunk 2]
```diff
@@ -149,26 +148,31 @@ def main(argv: Sequence[str]):
...
-    
     out_file_src_dec = open(data_path + 'it_dec_test.src', 'w')
     out_file_tgt_dec = open(data_path + 'it_dec_test.tgt', 'w')
+    ops_file = open(data_path + 'it_dec_test.ops', 'w')
```
Comment (santiaontanon-google) [inline datasets/pcfg/data_generation.py]: Similarly, "ops_file" is only for test, so maybe call it "out_file_ops_test", to match with the other naming scheme?

File: model/train.py [hunk 3]
```diff
@@ -588,99 +599,122 @@ def encode_tokens(labels):
...
-
-      # Save a checkpoint on one host after every checkpoint_freq steps.
-      save_checkpoint = (step % config.checkpoint_every_steps == 0 or
-                         is_last_step)
-      if config.save_checkpoints and save_checkpoint and jax.process_index() == 0:
-        with report_progress.timed("checkpoint"):
-          checkpoints.save_checkpoint(workdir, jax_utils.unreplicate(optimizer),
-                                      step)
+  if config.just_do_pred:
+      exemplars, test_acc, predicted_list = decode_and_calculate_acc(
```
Comment (santiaontanon-google) [inline model/train.py:609]: Maybe it's the Github visualization, but there some missing indentation here, no?

**Review by santiaontanon-google — APPROVED** [review]
Thanks Luana, just a few small nits

File: datasets/pcfg/data_generation.py [hunk 4]
```diff
@@ -118,10 +117,10 @@ def generate_intermediate_tasks(source_line, target_line):
   last_task = intermediate_tasks.pop()
   intermediate_tasks.append(last_task + ' END')
 
-  return intermediate_tasks
+  return intermediate_tasks, len(operations)
```
Comment (jtainslie) [inline datasets/pcfg/data_generation.py:121]: NIt: We should probably update the "Returns:" section of the docstring to mention `len(operations)`.

File: model/input_pipeline.py [hunk 5]
```diff
@@ -285,18 +285,20 @@ def filter_fn(x):
   dataset = dataset.repeat(num_epochs)
 
   if pack_examples:
-    dataset = pack_dataset(dataset, max_length)
+    dataset = pack_dataset(dataset, max_length, keys = ["inputs", "targets"])
```
Comment (jtainslie) [inline model/input_pipeline.py]: Nit: single quotes to be consistent in this file.
```suggestion
    dataset = pack_dataset(dataset, max_length, keys = ['inputs', 'targets'])
```

File: model/train.py [hunk 6]
```diff
@@ -40,6 +40,7 @@
 import jax.numpy as jnp
 import ml_collections
 import numpy as np
+import pickle as pkl
```
Comment (jtainslie) [inline model/train.py]: Nit: This should appear with the standard library imports.  (https://google.github.io/styleguide/pyguide.html#313-imports-formatting)

File: model/train.py [hunk 7]
```diff
@@ -342,7 +343,9 @@ def tokens_ids_to_logits(flat_ids, flat_cache):
 def pad_examples(x, desired_batch_size):
   """Expands batch to desired size by repeating last slice."""
   batch_pad = desired_batch_size - x.shape[0]
-  return np.concatenate([x, np.tile(x[-1], (batch_pad, 1))], axis=0)
+  if len(x.shape) == 1:
+      x = np.expand_dims(x, 1)
+  return np.concatenate([x, np.tile(x[-1], (batch_pad,1))], axis=0)
```
Comment (jtainslie) [inline model/train.py]: Formatting nit:
```suggestion
  return np.concatenate([x, np.tile(x[-1], (batch_pad, 1))], axis=0)
```

File: model/train.py [hunk 8]
```diff
@@ -415,6 +420,8 @@ def decode_and_calculate_acc(*, p_pred_step, p_init_cache, target, config,
     predicted = p_pred_step(pred_batch['inputs'], target, cache, decode.EOS_ID,
                             max_predict_length)
     predicted_list = [tohost(predicted)]
+    if use_annotations == True:
```
Comment (jtainslie) [inline model/train.py]: If `use_annotations` is expect to be boolean, then this would be simpler:
```suggestion
    if use_annotations:
```

File: model/train.py [hunk 9]
```diff
@@ -435,18 +442,22 @@ def decode_and_calculate_acc(*, p_pred_step, p_init_cache, target, config,
             predictions.append(decode_tokens(predicted[i]))
           if j == max_predict_loops - 1 and stop == False:
```
Comment (jtainslie) [inline model/train.py]: Nit: I should have caught this earlier, but it would be simpler to use:
```suggestion
          if j == max_predict_loops - 1 and not stop:
```
(Same with the other `stop == False` line above.)

**Review by jtainslie — APPROVED** [review]

File: model/train.py (hunk 3 above)
Comment (luanaruiz9) [reply model/train.py:609]: It looks like it was the Github visualization!
